src/core/utilities/cellmap.hpp

```diff
@@ -0,0 +1,566 @@
+                                                                      std::size_t itemIndex)
+{
+    bucketsLists_[cell].add(itemIndex);
+}
+
+
+template<std::size_t dim, std::size_t bucket_size, typename cell_index_t, typename key_t>
+template<typename Array, typename CellExtractor, typename>
+inline void CellMap<dim, bucket_size, cell_index_t, key_t>::add(Array const& items,
+                                                                CellExtractor extract)
+{
+    PHARE_LOG_SCOPE("CellMap::add (array)");
+    for (std::size_t itemIndex = 0; itemIndex < items.size(); ++itemIndex)
+    {
+        addToCell(extract(items[itemIndex]), itemIndex);
+    }
+}
+
+template<std::size_t dim, std::size_t bucket_size, typename cell_index_t, typename key_t>
+template<typename Array, typename CellExtractor, typename>
+inline void CellMap<dim, bucket_size, cell_index_t, key_t>::add(Array const& items,
+                                                                std::size_t first, std::size_t last,
+                                                                CellExtractor extract)
+
+{
+    PHARE_LOG_SCOPE("CellMap::add (iterator)");
+    for (auto itemIndex = first; itemIndex <= last; ++itemIndex)
+    {
+        addToCell(extract(items[itemIndex]), itemIndex);
+    }
+}
+
+
+template<std::size_t dim, std::size_t bucket_size, typename cell_index_t, typename key_t>
+template<typename Array, typename CellExtractor, typename>
+inline void CellMap<dim, bucket_size, cell_index_t, key_t>::add(Array const& items,
+                                                                std::size_t itemIndex,
+                                                                CellExtractor extract)
+{
+    PHARE_LOG_SCOPE("CellMap::add (add 1 index)");
+    addToCell(extract(items[itemIndex]), itemIndex);
+}
+
+
+
+template<std::size_t dim, std::size_t bucket_size, typename cell_index_t, typename key_t>
+template<typename Array, typename CellExtractor, typename>
+inline void CellMap<dim, bucket_size, cell_index_t, key_t>::erase(Array const& items,
+                                                                  std::size_t itemIndex,
+                                                                  CellExtractor extract)
+{
+    auto blist = list_at(extract(items[itemIndex]));
+    if (blist)
+    {
+        blist->get().remove(itemIndex);
+    }
+    else
+    {
+        std::cout << "item cannot be found\n";
+    }
+}
+
+
+template<std::size_t dim, std::size_t bucket_size, typename cell_index_t, typename key_t>
+template<typename Cells>
+inline void CellMap<dim, bucket_size, cell_index_t, key_t>::reserve(Cells const& cells)
+{
+    for (auto const& cell : cells)
+    {
+        bucketsLists_.emplace(cell, bucketlist_t{});
+    }
+}
+
+
+template<std::size_t dim, std::size_t bucket_size, typename cell_index_t, typename key_t>
+inline auto CellMap<dim, bucket_size, cell_index_t, key_t>::size() const
+{
+    PHARE_LOG_SCOPE("CellMap::size()");
+    auto s = 0u;
```

Comment:
"0u" is typicall std::uint32_t, whereas "size()" is typically std::size_t

Reply:
`http://coliru.stacked-crooked.com/a/3dd3bd831a9bbb8b`

Reply:
changed